website/addons/s3/static/s3NodeSettings.js

```diff
@@ -0,0 +1,337 @@
+'use strict';
+
+var ko = require('knockout');
+require('knockout-punches');
+require('knockout-mapping');
+var $ = require('jquery');
+var bootbox = require('bootbox');
+var Raven = require('raven-js');
+
+var $osf = require('osfHelpers');
+
+ko.punches.enableAll();
+
+var ViewModel = function(url, selector) {
+    var self = this;
+
+    self.url = url;
+    self.selector = selector;
+
+    self.nodeHasAuth = ko.observable(false);
+    self.userHasAuth = ko.observable(false);
+    self.userIsOwner = ko.observable(false);
+    self.ownerName = ko.observable('');
+
+    self.urls = ko.observable({});
+    self.loadedSettings = ko.observable(false);
+    self.bucketList = ko.observableArray([]);
+    self.loadedBucketList = ko.observable(false);
+    self.currentBucket = ko.observable('');
+    self.selectedBucket = ko.observable('');
+
+    self.accessKey = ko.observable('');
+    self.secretKey = ko.observable('');
+
+    self.loading = ko.observable(false);
+    self.creating = ko.observable(false);
+    self.creatingCredentials = ko.observable(false);
+
+    self.message = ko.observable('');
+    self.messageClass = ko.observable('text-info');
+
+    self.showSelect = ko.observable(false);
+
+    self.showSettings = ko.pureComputed(function() {
+        return self.nodeHasAuth();
+    });
+    self.disableSettings = ko.pureComputed(function() {
+        return !(self.userHasAuth() && self.userIsOwner());
+    });
+    self.showNewBucket = ko.pureComputed(function() {
+        return self.userHasAuth() && self.userIsOwner();
+    });
+    self.showImport = ko.pureComputed(function() {
+        return self.userHasAuth() && !self.nodeHasAuth();
+    });
+    self.showCreateCredentials = ko.pureComputed(function() {
+        return !self.nodeHasAuth() && !self.userHasAuth();
+    });
+    self.canChange = ko.pureComputed(function() {
+        return self.userIsOwner() && self.nodeHasAuth();
+    });
+    self.allowSelectBucket = ko.pureComputed(function() {
+        return (self.bucketList().length > 0 || self.loadedBucketList()) && (!self.loading());
+    });
+
+    self.fetchFromServer();
+};
+
+ViewModel.prototype.toggleSelect = function() {
+    this.showSelect(!this.showSelect());
+    if (!this.loadedBucketList()) {
+        this.fetchBucketList();
+    }
+};
+
+ViewModel.prototype.selectBucket = function() {
+    var self = this;
+    self.loading(true);
+    $osf.postJSON(
```

Comment:
For methods that send HTTP requests, it's nice to return the jqXHR, so that the promise can be used by the caller.